components/script/dom/bindings/str.rs

```diff
@@ -113,6 +113,74 @@ pub fn is_token(s: &[u8]) -> bool {
     })
 }
 
+/// Returns whether the language is matched, as defined by
+/// [RFC 4647](https://tools.ietf.org/html/rfc4647#section-3.3.2).
+pub fn extended_filtering(tag: &str, range: &str) -> bool {
+    let range_tags: Vec<&str> = range.split(',').collect();
+    let mut is_match: bool = true;
+
+    for range_tag in range_tags {
+        // step 1
+        let css_tags: Vec<&str> = range_tag.split('\x2d').collect();
+        let elem_tags: Vec<&str> = tag.split('\x2d').collect();
+        if tag.eq_ignore_ascii_case(range_tag) || range_tag.eq_ignore_ascii_case("*") {
+            return true;
+        }
+
+        // step 2
+        // Note: [Level-4 spec](https://drafts.csswg.org/selectors/#lang-pseudo) check for wild card
+        if !(css_tags[0].eq_ignore_ascii_case(elem_tags[0]) || css_tags[0].eq_ignore_ascii_case("*")) {
+            is_match = false;
+            continue;
+        }
+
+        // step 3
+        let css_tag_size = css_tags.len();
+        let elem_tag_size = elem_tags.len();
+        let mut css_tag_index = 1;
+        let mut elem_tag_index = 1;
+
+        while css_tag_size > css_tag_index {
+            // step 3a
+            if css_tags[css_tag_index].eq_ignore_ascii_case("*") {
+                css_tag_index += 1;
+                continue;
+            }
+            // step 3b
+            if elem_tag_size <= elem_tag_index {
```

Comment:
Using the new iterator method I proposed, this can adequately be implemented by just calling [`next`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#tymethod.next) on `tag_iter` and checking if the return value is `None`.